### ISSUE TITLE
Breakpoint stops on wrong line

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#263]
I have the following deep-copy function:

```javascript
// clone() function
// Creates a deep copy of an object, preserving circular references.
// Arguments:
//     o: The object to clone.
// Returns:
//     The new, cloned object.
function clone(o)
{
	var clones = arguments.length >= 2 ? arguments[1] : [];
	if (typeof o === 'object' && o !== null) {
		for (var i = 0; i < clones.length; ++i) {
			if (o === clones[i].original) {
				return clones[i].dolly;
			}
		}
		var dolly = o instanceof Array ? []
			: 'clone' in o && typeof o.clone === 'function' ? o.clone()
			: {};
		clones.push({ original: o, dolly: dolly });
		if (o instanceof Array || !('clone' in o) || typeof o.clone !== 'function') {
			for (var p in o) {
				dolly[p] = clone(o[p], clones);
			}
		}
		return dolly;
	} else {
		return o;
	}
}
```

When I set a breakpoint on `return dolly;`, Duktape will break on that line, but it will also break on `return o;` for which no breakpoint is set.

**Comment by svaarala** [comment]
There are some known line number issues; the main cause is noted in future work: https://github.com/svaarala/duktape/blob/master/doc/debugger.rst#improve-compiler-line-number-accuracy

Not sure if this is the case here though.

**Comment by fatcerberus** [comment]
I don't believe so; Duktape pauses at either return statement, but the client highlights the correct line in both cases, so Duktape definitely should be able to differentiate them for the purposes of triggering a breakpoint.

**Comment by svaarala** [comment]
This might actually just be caused by the logic used to decide when a breakpoint has been triggered. The algorithm is a bit more complex than one would initially think because a breakpoint may be set for a line which is never executed; a line may generate multiple opcodes during compilation; and the opcodes may not be in a linear sequence with respect to line numbers, see e.g.: https://github.com/svaarala/duktape/blob/master/doc/debugger.rst#line-transitions

Anyway, long story short, Duktape decides that a breakpoint has been hit when execution transitions to, or over, the target line. In this case if execution goes to the latter return, the return in the middle is "transitioned" in this way. The transition decision doesn't take into account where the code resides in the code block structure which is simply because that information is not available at execution time with the current compiler design.

**Comment by fatcerberus** [comment]
So if I understand you correctly, the condition to trigger a breakpoint is basically this, then?
`if (cur_line >= breakpoint_line) pause()`

That seems unintuitive.  And in any case, if a breakpoint is set on a line that isn't executed, I wouldn't expect it to be hit at all.  Unless I misunderstand?

**Comment by svaarala** [comment]
At some point breakpoints were triggered only by code reaching that exact line. Unfortunately this approach had some shortcomings so the transition based approach was used instead.

But perhaps a better fix would be to move back to breakpoints being triggered only when execution transition to exactly that line, and then changing AddBreak so that it will either reject an attempt to add a breakpoint on a line with no emitted code, or so that it would find the next line with emitted code.

**Comment by svaarala** [comment]
This is the current logic: https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1564-L1568

Regarding "breakpoint is set on a line that isn't executed", from a practical perspective a bytecode-oriented VM doesn't execute "lines" of code. It just executes opcodes, and breakpoint conditions need to ultimately be formulated in terms of these.

The problem with stopping on an exact line is that some lines are *never* executed - unless opcodes are emitted such a line can just be skipped entirely. If a breakpoint was set there and is never triggered, that'd be confusing and was one of the reasons the transition based approach was used.

But I agree the current behavior isn't very good either but I'm not sure what the best fix is, taking into account the information available at runtime. Like I said above, perhaps exact breakpoints and a small AddBreak modification would do the trick.

**Comment by svaarala** [comment]
Ideally the condition could be: the line has been transitioned *and* the statement sequence where the breakpoint belongs (e.g. the "then" clause of an if statement) has not been skipped. But that information is not available at runtime right now.

**Comment by svaarala** [comment]
Ah, now I remember why the AddBreak change (to use the "next" line when the specified line has no code) isn't trivial: there are sometimes several possible "next lines". Consider for example:

```js
switch (x) {
    // try to set breakpoint here
    case 1: print('foo'); break;
    case 2: print('bar'); break;
    case 3: print('quux'); break;
}
```

If you tried to AddBreak on the line above, you'd need three separate exact breakpoints to cover the possible execution paths after the line (plus one more for "nothing matched" I guess).

The concrete problem is formulated in terms of bytecode though: if the specified line has no emitted opcodes, how do you find the next relevant line when you're only given bytecode. One would maybe need to follow JUMP opcodes etc to correctly answer that. It might be just easier and "good enough" to reject an AddBreak for a line that contains no emitted opcodes.

**Comment by fatcerberus** [comment]
For what it's worth, MSVC simply won't allow you to set a breakpoint on a non-executable line (such as a variable declaration).  In the debugger, this is disallowed outright, and if you attempt to do so at design time, the breakpoint will be disabled once you go to run the program.

**Comment by svaarala** [comment]
Some other debuggers do allow those non-aligned breakpoints though, and that's probably what I'd expect. Though it wouldn't be bad if the debug UI gave a clear indication what failed and perhaps suggested another line.

If you'd like to test how this approach worked, you could just quickly change:

```
	if (act->prev_line < bp->line && line >= bp->line) {
```

to

```
	if (act->prev_line < bp->line && line == bp->line) {
```

in the combined `duktape.c` and see what happens.

**Comment by fatcerberus** [comment]
I'll try this later.  In theory the only potential issue is if Duktape assigns the wrong line number to a given instruction, the breakpoint will be missed.  In practice I haven't seen this though--when single-stepping it reliable moves from one line to the next without skipping any or scrambling them.

**Comment by svaarala** [comment]
Ok - let me know what happens :) The current compiler does have some known issues with lines not being optimally assigned to opcodes. For instance, IIRC there is a case where some opcodes are assigned to the closing brace of a block (maybe it was for some loop case, not sure). But it shouldn't be a major issue.

**Comment by svaarala** [comment]
Changing the behavior for AddBreak like suggested above (to reject breakpoint on a line with no code) is actually not a small change, because currently breakpoints can be added in advance, and are rechecked based on fileName and a line number range whenever the function changes. This is actually a feature because it's nice to be able to add a breakpoint for a module not yet loaded for example.

**Comment by fatcerberus** [comment]
Yeah, I definitely understand the lazy breakpoint handling, I would probably implement it the same way.  I just didn't expect a breakpoint set in one scope to pause in a different one, that caught me off guard. :)

**Comment by fatcerberus** [comment]
So I ended up changing the condition to the following:
`if (act->prev_line != bp->line && line == bp->line) {`

This seems to work well in my case at least, and also catches backward movement as well (in a loop, for example).

**Comment by svaarala** [comment]
That might be the best of the viable short term alternatives, here's a branch with that change included: https://github.com/svaarala/duktape/compare/breakpoint-line-rule-gh263

I'll merge the change in a few days unless there are better ideas for a change in 1.3 release.

**Comment by fatcerberus** [comment]
Nice, thanks.

I definitely understand the difficulties here; breakpoints are described in terms of source while actual execution commences in the PC domain.  I am certainly familiar with the tedium of working across domains like this, it's not always easy.

That said, Duktape's source map has proven to be pretty accurate; with the exception of that one issue a while back with `Duktape.act()` (which turned out to be a bug anyway), I've found reported line numbers to be faithful 99% of the time.

Maybe just make a note in the guide that breakpoints set on non-executable lines will no longer be hit as of 1.3.0, to avoid surprises. :)

**Comment by svaarala** [comment]
The branch has a RELEASE.rst entry for it, but I agree it should be mentioned in the release notes file. Added a note to #203 for that.

**Comment by svaarala** [comment]
Updated internal debugger doc a bit, now merged to master.

**Comment by fatcerberus** [comment]
Thanks. :)

Just a bit of clarification on why this change is so important to me: I will often set a breakpoint inside a conditional block once I've determined the circumstances under which a bug manifests, to view the state of the program at the time the triggering condition is met.  It's a surprisingly effective debugging technique--particularly when the offending condition is inside of a tight loop (making single-stepping too tedious)--but it rather defeats the purpose when Duktape proceeds to break anyway on the line *after* the condition. :stuck_out_tongue_winking_eye:

**Comment by svaarala** [comment]
Yeah, I don't think you'll get much argument that the previous behavior was confusing.

The pain here was to choose between somewhat imperfect alternatives. The perfect alternative would require internal changes so that we'd have a bit more information about code block structure etc.
